src/java/org/apache/cassandra/db/memtable/AbstractMemtable.java

```diff
@@ -94,6 +99,48 @@ protected void updateMin(AtomicLong minTracker, long newValue)
         }
     }
 
+    private final void updateMinPartitionKey(DecoratedKey key)
+    {
+        while (true)
+        {
+            DecoratedKey currentMinKey = minPartitionKey.get();
+            if (currentMinKey != null && key.compareTo(currentMinKey) >= 0)
+                break;
+            if (minPartitionKey.compareAndSet(currentMinKey, key))
+                break;
+        }
+    }
+
+    private void updateMaxPartitionKey(DecoratedKey key)
+    {
+        while (true)
+        {
+            DecoratedKey currentMaxKey = maxPartitionKey.get();
+            if (currentMaxKey != null && key.compareTo(currentMaxKey) <= 0)
+                break;
+            if (maxPartitionKey.compareAndSet(currentMaxKey, key))
+                break;
+        }
+    }
+
+    protected final void updateMinMaxPartitionKey(DecoratedKey key)
+    {
+        updateMinPartitionKey(key);
```

Comment:
This could cause non-trivial overhead under contention. Have you run some benchmark (e.g. `WriteBench`) to validate it's okay?

Alternatively, with `SkipListMemtable` you can get the smallest and largest key by simply querying the map. With `TrieMemtable` you can still get the smallest key using iteration, and you can store the largest individually for each shard to avoid the CAS loop (replaced with iteration once reversed iteration is implemented).

As a legacy secondary index populated with growing keys is going to be the worst case for this, I would at least implement the methods directly from `partitions.firstKey()` and `partitions.lastKey()` for `SkipListMemtable`.